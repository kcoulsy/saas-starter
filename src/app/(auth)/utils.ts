import { pageRoutes } from '@src/constants/routes';
import { authOptions } from '@src/pages/api/auth/[...nextauth]';
import { getServerSession } from 'next-auth';
import { redirect } from 'next/navigation';

export async function redirectIfLoggedIn() {
  const session = await getServerSession(authOptions);

  if (session?.user?.email) {
    redirect(pageRoutes.dashboard);
  }
}
