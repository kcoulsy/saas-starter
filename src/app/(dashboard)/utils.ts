import { getServerSession } from 'next-auth';
import { redirect } from 'next/navigation';
import { pageRoutes } from '@src/constants/routes';
import { authOptions } from '@src/server/auth';

export async function redirectIfNotLoggedIn() {
  const session = await getServerSession(authOptions);

  if (!session?.user?.email) {
    redirect(pageRoutes.login);
  }
}
