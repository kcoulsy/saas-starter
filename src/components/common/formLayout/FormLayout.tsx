import React, { HTMLInputTypeAttribute } from 'react';
import { UseFormRegisterReturn } from 'react-hook-form';
import Button from '../button/Button';
import FormInput from '../formInput/FormInput';

export interface FormLayoutProps {
  title: string;
  description: string | React.ReactNode;
  formFields: {
    id: string;
    type: HTMLInputTypeAttribute;
    label: string;
    placeholder: string;
    register: UseFormRegisterReturn<string>;
    errors: string[];
  }[];
  footer?: string | React.ReactNode;
  submitButtonLabel: string;
  errors: string[];
  isLoading?: boolean;
}

const FormLayout = ({
  title,
  description,
  formFields,
  submitButtonLabel,
  footer,
  errors,
  isLoading = false,
}: FormLayoutProps) => {
  return (
    <div className="">
      <div className="">
        <h1 aria-level={1} aria-label={title} className="text-2xl font-extrabold leading-6 text-gray-800">
          {title}
        </h1>
        <h2 className="text-sm mt-4 font-medium leading-none text-gray-500">{description}</h2>
        <div className="mt-4 w-full grid gap-y-4 ">
          {formFields.map((formInputProps) => (
            <FormInput key={formInputProps.id} {...formInputProps} />
          ))}
        </div>
        {errors && (
          <ul>
            {errors.map((error) => (
              <li key={error} className="mt-1">
                <small className="text-red-500">{error}</small>
              </li>
            ))}
          </ul>
        )}
        {footer && <div className="mt-4  text-sm font-medium leading-none text-gray-500">{footer}</div>}
        <Button label={submitButtonLabel} type="submit" classes="mt-4" isLoading={isLoading} />
      </div>
    </div>
  );
};

export default FormLayout;
