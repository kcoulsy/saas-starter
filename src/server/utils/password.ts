import crypto from 'crypto';

export async function hash(password: string) {
  return new Promise<string>((resolve, reject) => {
    // generate random 16 bytes long salt
    const salt = crypto.randomBytes(16).toString('hex');

    crypto.scrypt(password, salt, 64, (err, derivedKey) => {
      if (err) reject(err);
      resolve(`${salt}:${derivedKey.toString('hex')}`);
    });
  });
}

export async function verify(password: string, hashString: string) {
  return new Promise((resolve, reject) => {
    const [salt, key] = hashString.split(':');
    // Mitigating timing attacks
    // https://dev.to/bdougherty/comment/1c1fa
    const keyBuffer = Buffer.from(key || '', 'hex');
    crypto.scrypt(password, salt || '', 64, (err, derivedKey) => {
      if (err) reject(err);
      resolve(crypto.timingSafeEqual(keyBuffer, derivedKey));
    });
  });
}
