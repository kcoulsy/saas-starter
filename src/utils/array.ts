/**
 * Helper method for array filtering to maintain type safety
 * rather than using .filter(Boolean)
 *
 * @param value unknown value
 * @returns boolean
 */
export function notEmpty<TValue>(value: TValue | null | undefined): value is TValue {
  return value !== null && value !== undefined;
}
