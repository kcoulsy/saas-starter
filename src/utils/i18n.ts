import { initReactI18next } from 'react-i18next';
import i18n from 'i18next';
import enCommonTranslations from '@src/locales/en/common.json';
import enLoginFormTranslations from '@src/locales/en/login-form.json';
import enRegisterFormTranslations from '@src/locales/en/register-form.json';

i18n.use(initReactI18next).init({
  lng: 'en',
  fallbackLng: 'en',

  // have a common namespace used around the full app
  ns: ['translations'],
  defaultNS: 'translations',

  interpolation: {
    escapeValue: false, // not needed for react!!
  },

  debug: true,

  resources: {
    en: {
      common: enCommonTranslations,
      'login-form': enLoginFormTranslations,
      'register-form': enRegisterFormTranslations,
    },
  },
});

export default i18n;
