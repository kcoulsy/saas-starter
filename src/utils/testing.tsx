import React from 'react';
import { render } from '@testing-library/react';
// import TypesafeI18n from '@src/i18n/i18n-react';
// import { loadAllLocales } from '@src/i18n/i18n-util.sync';
import Providers from '@src/app/providers';

// const wrapper = ({ children }: { children: React.ReactNode }) => {
//   // loadAllLocales();
//   return <TypesafeI18n locale="en">{children}</TypesafeI18n>;
// };

const renderComponent = (component: React.ReactElement) => {
  return render(component, { wrapper: Providers });
};

export { renderComponent };
