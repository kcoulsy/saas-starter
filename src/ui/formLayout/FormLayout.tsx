import React from 'react';
import { cn } from '@src/utils/cn';
import Button from '../button/Button';
import FormInput, { FormInputProps } from '../formInput/FormInput';

export interface FormLayoutProps {
  title?: string;
  description?: string | React.ReactNode;
  formFields: FormInputProps[];
  footer?: string | React.ReactNode;
  submitButtonLabel?: string;
  errors: string[];
  isLoading?: boolean;
  buttonClasses?: string;
  successMessage?: string;
}

const FormLayout = ({
  title,
  description,
  formFields,
  submitButtonLabel,
  footer,
  errors,
  isLoading = false,
  buttonClasses,
  successMessage,
}: FormLayoutProps) => {
  return (
    <div className="">
      <div className="">
        {title && (
          <h1 aria-level={1} aria-label={title} className="text-2xl font-extrabold leading-6 text-gray-800">
            {title}
          </h1>
        )}
        {description ? <h2 className="text-sm mt-4 font-medium leading-none text-gray-500">{description}</h2> : null}
        <div className="mt-4 w-full grid gap-y-4 ">
          {formFields.map((formInputProps) => (
            <FormInput key={formInputProps.id} {...formInputProps} />
          ))}
        </div>
        {errors && (
          <ul>
            {errors.map((error) => (
              <li key={error} className="mt-1">
                <small className="text-red-500">{error}</small>
              </li>
            ))}
          </ul>
        )}
        {footer && <div className="mt-4  text-sm font-medium leading-none text-gray-500">{footer}</div>}
        {submitButtonLabel && (
          <Button label={submitButtonLabel} type="submit" classes={cn('mt-4', buttonClasses)} isLoading={isLoading} />
        )}
        {successMessage && <div className="mt-4">{successMessage}</div>}
      </div>
    </div>
  );
};

export default FormLayout;
