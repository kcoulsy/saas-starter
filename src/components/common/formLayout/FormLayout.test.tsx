import { renderComponent } from '@src/utils/testing';
import { composeStories } from '@storybook/react';
import { screen } from '@testing-library/dom';
import { describe } from 'vitest';
import * as stories from './FormLayout.stories';

const { Default, WithErrors } = composeStories(stories);

describe('FormLayout', () => {
  it('should render correctly', () => {
    const { asFragment } = renderComponent(<Default />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with errors', () => {
    const { asFragment } = renderComponent(<WithErrors />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the title', () => {
    renderComponent(<Default title="Test Title" />);
    expect(screen.getByText('Test Title')).toBeInTheDocument();
  });

  it('should render the description', () => {
    renderComponent(<Default description="Test Description" />);
    expect(screen.getByText('Test Description')).toBeInTheDocument();
  });

  it('should allow jsx in description', () => {
    renderComponent(<Default description={<span data-testid="jsx-element">Test Description</span>} />);
    expect(screen.getByTestId('jsx-element')).toBeInTheDocument();
  });

  it('should render all the form fields', () => {
    const mockFormFields = [
      {
        id: 'id',
        type: 'text',
        label: 'Label',
        placeholder: 'Placeholder',
        register: {
          name: 'name',
          onBlur: async () => alert('onBlur'),
          onChange: async () => alert('onChange'),
          ref: () => false,
        },
        errors: [],
      },
      {
        id: 'id2',
        type: 'text',
        label: 'Label',
        placeholder: 'Placeholder',
        register: {
          name: 'name',
          onBlur: async () => alert('onBlur'),
          onChange: async () => alert('onChange'),
          ref: () => false,
        },
        errors: [],
      },
      {
        id: 'id3',
        type: 'text',
        label: 'Label',
        placeholder: 'Placeholder',
        register: {
          name: 'name',
          onBlur: async () => alert('onBlur'),
          onChange: async () => alert('onChange'),
          ref: () => false,
        },
        errors: [],
      },
    ];

    renderComponent(<Default formFields={mockFormFields} />);

    expect(screen.getAllByRole('textbox')).toHaveLength(mockFormFields.length);
  });

  it('should render all the errors', () => {
    const mockErrors = ['Error 1', 'Error 2', 'Error 3'];

    renderComponent(<WithErrors errors={mockErrors} />);

    expect(screen.getByText('Error 1')).toBeInTheDocument();
    expect(screen.getByText('Error 2')).toBeInTheDocument();
    expect(screen.getByText('Error 3')).toBeInTheDocument();
  });

  it('should render the submit button', () => {
    renderComponent(<Default submitButtonLabel="Test Submit" />);
    expect(screen.getByText('Test Submit')).toBeInTheDocument();
  });
});
