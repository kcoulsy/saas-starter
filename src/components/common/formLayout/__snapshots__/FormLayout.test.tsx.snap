// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormLayout > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="lg:w-1/3 md:w-1/2 w-full"
  >
    <div
      class=""
    >
      <div
        class=""
      >
        <h1
          aria-label="Example Form"
          aria-level="1"
          class="text-2xl font-extrabold leading-6 text-gray-800"
        >
          Example Form
        </h1>
        <h2
          class="text-sm mt-4 font-medium leading-none text-gray-500"
        >
          This is an example form
        </h2>
        <div
          class="mt-4 w-full grid gap-y-4 "
        >
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            />
          </div>
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id2"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id2"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            />
          </div>
        </div>
        <ul />
        <button
          aria-label="Submit"
          class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-indigo-700 border rounded hover:bg-indigo-600 focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700 mt-4"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
  ;
</DocumentFragment>
`;

exports[`FormLayout > should render correctly with errors 1`] = `
<DocumentFragment>
  <div
    class="lg:w-1/3 md:w-1/2 w-full"
  >
    <div
      class=""
    >
      <div
        class=""
      >
        <h1
          aria-label="Form With Errors"
          aria-level="1"
          class="text-2xl font-extrabold leading-6 text-gray-800"
        >
          Form With Errors
        </h1>
        <h2
          class="text-sm mt-4 font-medium leading-none text-gray-500"
        >
          This form will include some errors
        </h2>
        <div
          class="mt-4 w-full grid gap-y-4 "
        >
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            >
              <li>
                <small
                  class="text-red-500"
                >
                  Input Error 1
                </small>
              </li>
              <li>
                <small
                  class="text-red-500"
                >
                  Input Error 2
                </small>
              </li>
            </ul>
          </div>
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id2"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id2"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            >
              <li>
                <small
                  class="text-red-500"
                >
                  Input Error 1
                </small>
              </li>
              <li>
                <small
                  class="text-red-500"
                >
                  Input Error 2
                </small>
              </li>
            </ul>
          </div>
        </div>
        <ul>
          <li
            class="mt-1"
          >
            <small
              class="text-red-500"
            >
              Form Error 1
            </small>
          </li>
          <li
            class="mt-1"
          >
            <small
              class="text-red-500"
            >
              Form Error 2
            </small>
          </li>
        </ul>
        <button
          aria-label="Submit"
          class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-indigo-700 border rounded hover:bg-indigo-600 focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700 mt-4"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
  ;
</DocumentFragment>
`;

exports[`FormLayout > should render correctly with footer 1`] = `
<DocumentFragment>
  <div
    class="lg:w-1/3 md:w-1/2 w-full"
  >
    <div
      class=""
    >
      <div
        class=""
      >
        <h1
          aria-label="Example Form"
          aria-level="1"
          class="text-2xl font-extrabold leading-6 text-gray-800"
        >
          Example Form
        </h1>
        <h2
          class="text-sm mt-4 font-medium leading-none text-gray-500"
        >
          This is an example form
        </h2>
        <div
          class="mt-4 w-full grid gap-y-4 "
        >
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            />
          </div>
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id2"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id2"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            />
          </div>
        </div>
        <ul />
        <div
          class="mt-4  text-sm font-medium leading-none text-gray-500"
        >
          <div>
            Footer here with a link 
            <a
              class="underline"
              href="#"
            >
              click me
            </a>
          </div>
        </div>
        <button
          aria-label="Submit"
          class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-indigo-700 border rounded hover:bg-indigo-600 focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700 mt-4"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
  ;
</DocumentFragment>
`;

exports[`FormLayout > should render correctly with link in description 1`] = `
<DocumentFragment>
  <div
    class="lg:w-1/3 md:w-1/2 w-full"
  >
    <div
      class=""
    >
      <div
        class=""
      >
        <h1
          aria-label="Example Form"
          aria-level="1"
          class="text-2xl font-extrabold leading-6 text-gray-800"
        >
          Example Form
        </h1>
        <h2
          class="text-sm mt-4 font-medium leading-none text-gray-500"
        >
          <div>
            hello 
            <a
              class="underline"
              href="#"
            >
              world
            </a>
          </div>
        </h2>
        <div
          class="mt-4 w-full grid gap-y-4 "
        >
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            />
          </div>
          <div>
            <div
              class="relative flex align-middle"
            >
              <label
                class="text-sm font-medium leading-none text-gray-800 capitalize relative flex-1"
                for="id2"
              >
                Label
                <input
                  aria-label="Label"
                  class="bg-gray-200 border rounded focus:outline-none text-xs font-medium leading-none text-gray-800 py-3 w-full pl-3 mt-2"
                  id="id2"
                  name="name"
                  placeholder="Placeholder"
                  type="text"
                />
              </label>
            </div>
            <ul
              class="mt-1"
            />
          </div>
        </div>
        <ul />
        <button
          aria-label="Submit"
          class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-indigo-700 border rounded hover:bg-indigo-600 focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700 mt-4"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
  ;
</DocumentFragment>
`;
